### ISSUE TITLE
Choose directory to search for .editorconfig or package.json

### ISSUE THREAD

**Issue by jonathantneal** [ahmadnassri/echint#60]
Right now, I can specify the path to an `.editorconfig` file, or I can specify a node module using `extends`.

If unspecified, `.editorconfig` looks within the current working directory. Similarly, if `extends` is used, it looks within the current directory.

However, I’m using echint to check a directory that is not the current directory. This means I need it to sniff within that directory for either the `.editorconfig` or the node module.

This gets tricker when we consider that `pkg-config` may find land us in yet another directory.

How do you think we should handle this? Should `pkg-config` also return the current directory it found the `package.json` in? Or should we patch this with a `cwd` option that lets you specify the directory to work from (which would fallback to `process.cwd()`, as is used in the code)? The later option will be easier to implement immediately. However, it wouldn’t resolve the situation where an ancestor `package.json` found the `echint` config, and then attempted to look for that package from the `cwd` (which is a problem we face with the current implementation already).

**Comment by jonathantneal** [comment]
Here are some methods that could be used to achieve this effect.

A no-dependency method to return the JSON of a file.

```js
// promise read file
const readJSON = (file) => new Promise(
	(resolve, reject) => fs.readFile(
		file,
		'utf8',
		(error, data) => error ? reject(error) : resolve(JSON.parse(data))
	)
)
```

A no-dependency method to return the nearest path to and contents of a `package.json`.

```js
// promise nearest package
const readNearestPackage = (cwd = process.cwd()) => readJSON(
	// path to package
	path.join(cwd, 'package.json')
).then(
	// path to and content of package
	(result) => ({
		cwd,
		pkg: result
	}),
	// thrown error or nested search
	(error) => {
		if (error.code === 'ENOENT' && path.dirname(cwd) !== cwd) {
			// search deeper if package is not found and a parent directory exists
			return readNearestPackage(
				path.dirname(cwd)
			);
		} else {
			// otherwise, throw an error
			throw 'Could not load package'
		}
	}
);
```

A simple, no-dependency method to return the `echint` config from the nearest `package.json`.

```js
// promise echint config from the nearest package
const readNearestEcHintConfig = (cwd = process.cwd()) => readNearestPackage(cwd).then(
	// current directory and echint config from package
	(result) => ({
		config: result.pkg.echint || {},
		cwd: result.cwd
	})
)
```

From here, `getModuleEditorConfig` could be used to get the nearest package and `echint` configuration. If that config includes an `extends` property, we can use that and the `cwd` to get the sanitized config module name with `extends.replace(/^(echint-config-)?/, 'echint-config-')`, and the path to the module with `path.resolve(cwd, 'node_modules', packageName)`.

With the path to the module, we can either sniff its `package.json` for `main`, or use `.editorconfig`, and now we have the path to our module’s editorconfig file.

**Comment by ahmadnassri** [comment]
I believe this is on the right path, happy to take PRs, or can look into making those changes later myself (~48 hours)

**Comment by stefanmaric** [comment]
Just to add references: this issue is related to issue #59 and being partially addressed by PR #80
